### DISCUSSION TITLE
Using python-dali as a slave, not a master

### DISCUSSION THREAD

**Discussion by hollymcr** [sde1000/python-dali#126] [General]
Is it possible to use python-dali as a slave, rather than a master?

That is, could I use it to mimic one or more light fittings, connected to a real DALI line, responding to DALI commands and status queries from a real DALI master?

What I would like is to create a test program that mimics (up to) all 64 DALI lamps on a line, for use when testing the configuration of a DALI gateway. It would allow me to mark any individual lamp as present/missing/failed/etc and show the light level that had been sent from the master. It would allow me to confirm that my gateway configuration does what I expect without rigging 64 lamps up on a bench.

(If not, are there similar libraries which might? Maybe there are libraries that python-dali uses for testing without hardware present?)

Issues likely isn't the right place for this but I couldn't see a mailing list or other option, apologies if I missed it,

**Comment by sde1000** [comment]
Interesting question!

It's certainly not what the library was intended to do, but it's probably not impossible.

There's a partial implementation of DALI gear used for testing the library without hardware present; it's in [dali/tests/fakes.py](https://github.com/sde1000/python-dali/blob/master/dali/tests/fakes.py). It is *not* a standard-compliant implementation!

You would have to hook it up to a driver for an interface that's able to listen for forward frames and send backward frames. I don't know which interfaces support this. I can't see any mention of being able to send backward frames in [the LUBA protocol documentation](https://www.lunatone.com/wp-content/uploads/2021/04/LUBA_Protocol_EN.pdf), for example.

**Reply by jktjkt** [reply]
I have not tried this myself, but the [LUBA documentation](https://www.lunatone.com/wp-content/uploads/2021/04/LUBA_Protocol_EN.pdf) says in section 2.1.1 on page 7 that it supports variable-length frames. This is supported by the [device datasheet](https://www.lunatone.com/wp-content/uploads/2023/01/24138215_DALI-2_USB30mA_EN_D0114.pdf) that says that the interface is capable of sending "backchannel" frames:

![image](https://github.com/sde1000/python-dali/assets/2631925/b295d974-88b2-48f6-bdef-e54193ba5bc8)

I do not have that device, I have *DALI USB HS – for DIN Rail* (Art. Nr.: 24138923-HS) which talks the undocumented (right?) Lunatone protocol over USB HID, which is supported by the `tridonic` HID driver. That one is also documented to support backward frames ([page 8](https://www.lunatone.com/wp-content/uploads/2018/03/24138923_DALI_USB_EN_D0052-1.pdf)):

![image](https://github.com/sde1000/python-dali/assets/2631925/009879f0-311e-427f-a9a4-d1bdd61cd410)

The driver currently hardcodes either 16-bit or 24-bit forward frames only, but given that there's already an enum for an 8-bit frame (`_SEND_MODE_DALI8`), I guess that this *might* actually work with a trivial patch. **If** the HW supports this, of course, which I don't know. In order to test this, one needs either:
- Some HW bus master that's capable of doing *something* upon seeing a particular backward frame. I have Lunatone DALI-2 CS (86458670-O-W16-AP) which can do that, but testing this will require some effort. I might do that :).
- Or another interface that's capable of sniffing the communication. I don't have a second one.

This project sounds like a wonderful idea. Personally, I do not need this real, HW implementation for actually sending backward frames to bus masters, but the state machine of control gear which implements, say, fading timers and DT8 logic in response to actual commands as reported by my interface' bus monitor would be awesome. If something like this is in place, I'll get a remote dashboard which tells me live status of my lights for free :). It's also plenty of work, of course.

**Reply by jktjkt** [reply]
That trivial patch is here: https://github.com/sde1000/python-dali/pull/127

**Reply by jktjkt** [reply]
...and it works as expected with the HW that I have. Now @hollymcr just needs to implement that "simulator of DALI control gear", and we'll be all set :).

**Reply by hollymcr** [reply]
What have I opened myself up to? :-)

The hardware I have (albeit not handy - it's in a locked drawer on site somewhere) is a Tridonic USB interface. My instinct if you're confident this should work from your tests on your Lunatone interface is to get one of those so we're on the same page - which interface is it that you have?

I can see two use cases for this. One is as I described - the ability to simulate in software all or part of a DALI line, including taking lamps out or failing them, etc, which would be useful when writing application software at the master end. The second would be as a way to take, say, a Raspberry Pi, and turn it into a DALI slave device so that it can be connected to a real DALI line and respond to DALI events (eg PIRs/switches). The latter case would also require a direct hardware interface of some kind (USB interfaces are too expensive for the sort of hobbyist projects that this might appeal to) but there do seem to be some DALI HATs for the Pi out there, although they also seem designed to be master not slave.

At this point this is something of personal interest to me but not something my employer will commit time to so I don't know if I'll find the time to make anything of it (or to justify buying a new DALI interface), so don't get your hopes up that this will lead anywhere, but I do want to see how far I can get.

**Reply by hollymcr** [reply]
I just (re)discovered the Mikroe Click board range: they have both DALI and DALI-2 Click boards for $25, and a USB interface which would allow them to be used on a PC for $29 - between them a much cheaper option that the USB->DALI hardware I've looked at in the past. If these can act as DALI slaves (I've asked) then that would be something I'd be very interested in pursuing. 

[USB Adapter](https://www.mikroe.com/click-usb-adapter)
[DALI Click](https://www.mikroe.com/dali-click)
[DALI-2 Click](https://www.mikroe.com/dali-2-click)

**Reply by sde1000** [reply]
The Tridonic DALI USB is a rebadged Lunatone device. It should work.

I use a DALI Click with a Pi Pico to generate random frames for testing. I wasn't aware there was a second-generation version, I'll have to check that out. The Pico just emits DALI frames using one of its PIO units — no bit-banging required. (I haven't needed to implement receive or multi-master support yet! That will be more tricky.)

There is no physical layer difference between the interfaces on DALI gear and DALI controllers, the difference is entirely in the firmware of the microcontroller.

The USB adapter looks like it will allow you to connect a Click board to USB, but it doesn't necessarily have any DALI protocol support itself — you would have to check with them. The Click boards are easy to interface to any microcontroller, anyway.

**Comment by TA1DB** [comment]
@sde1000 
> I use a DALI Click with a Pi Pico to generate random frames for testing

That's good, I am playing with Pi Pico PIO + Dali Click as well, just for training.. I would love to hear about your works.
